File: C++/005. Longest Palindromic Substring.cpp
#include <iostream>
#include <vector>

class Solution {
public:
    std::string longestPalindrome(std::string s) {
        auto n = int(s.size());
        int start = 0, end = 0;
        std::vector<std::vector<bool>> dp (unsigned(n), std::vector<bool>(unsigned(n), true));
        for (int j = 0; j < n; j++) {
            for (int i = j - 1; i >= 0; i--) {
                dp[i][j] = dp[i + 1][j - 1] && s[i] == s[j];
                if (dp[i][j] && end - start < j - i) {
                    start = i;
                    end   = j;
                }
            }
        }
        return s.substr(unsigned(start), unsigned(end - start + 1));
    }
};

File: Go/005. Longest Palindromic Substring.go
func longestPalindrome(s string) string {
	var dp [][]bool
	dp = make([][]bool, len(s), len(s))
	for i := 0; i < len(s); i++ {
		dp[i] = make([]bool, len(s), len(s))
		for j := 0; j < len(s); j++ {
			dp[i][j] = true
		}
	}
	start, end := 0, 0
	for j := 0; j < len(s); j++ {
		for i := j - 1; i >= 0; i-- {
			dp[i][j] = dp[i + 1][j - 1] && s[i] == s[j]
			if dp[i][j] && end - start < j - i {
				start, end = i, j
			}
		}
	}
	return s[start : end + 1]
}